cub/test/c2h/generators.cuh

```diff
@@ -110,5 +145,34 @@ template <typename T>
 thrust::device_vector<T>
 gen_uniform_offsets(seed_t seed, T total_elements, T min_segment_size, T max_segment_size);
 
-} // c2h
+/**
+ * @brief Generates key-segment ranges from an offsets-array like the one given by
+ * `gen_uniform_offset`.
+ */
+template <typename key_t, typename offset_t>
+thrust::device_vector<key_t>
+get_key_segments(const thrust::device_vector<offset_t> &segment_offsets, offset_t total_elements)
+{
+  // Constant iterator returning '1' for any index
+  key_t non_zero_constant{};
+  detail::init_non_zero_constant(non_zero_constant);
+  auto const_one_it = thrust::make_constant_iterator(non_zero_constant);
+
+  // Iterator of segment offsets (skipping the first segment offset of '0')
+  auto segment_offset_it = thrust::next(segment_offsets.cbegin());
+
+  // Prepare keys array, scattering '1' to segment offsets
+  thrust::device_vector<key_t> key_segments(total_elements);
+  auto num_segments = segment_offsets.size() - 1;
+  thrust::scatter_if(const_one_it,
+                     const_one_it + num_segments,
+                     segment_offset_it,
+                     segment_offset_it,
+                     key_segments.begin(),
+                     detail::le_comparator_op<offset_t>{total_elements - 1});
+
+  thrust::inclusive_scan(key_segments.cbegin(), key_segments.cend(), key_segments.begin());
+  return key_segments;
+}
```

Comment:
This will end up in each TU. Could you move it into generators.cu?

Reply:
Moved it to generators and based it on the run-length decode-based approach. Would be great if you could give it another quick check.